packages/berry-pnpify/sources/index.ts

```diff
@@ -0,0 +1,176 @@
+import fs from 'fs';
+import os from 'os';
+import path from 'path';
+
+function readdirNodeModules(pnp: any, dir: string, issuer: string): string[] | undefined {
+  const dirs: { [dir: string]: string[] } = {};
+  const locator = pnp.findPackageLocator(issuer);
+  if (!locator) {
+    return undefined;
+  }
+  const info = pnp.getPackageInformation(locator);
+  if (info.packageLocation + '/' !== issuer) {
+    return undefined;
+  }
+  const packages: string[] = Array.from(info.packageDependencies.keys());
+  packages.forEach((pkg: string) => {
+    dirs[''] = dirs[''] || [];
+    if (pkg[0] === '@') {
+      const [ scopeName, packageName ] = pkg.split('/');
+      dirs[scopeName] = dirs[scopeName] || [];
+      dirs[scopeName].push(packageName);
+      if (dirs[''].indexOf(scopeName) < 0) {
+        dirs[''].push(scopeName);
+      }
+    } else {
+      dirs[''].push(pkg);
+    }
+  });
+  return dirs[dir];
+};
+
+function getIssuerForNodeModulesPath(pnp: any, pathname: string) {
+  const locator = pnp.findPackageLocator(pathname);
+  if (!locator) {
+    return undefined;
+  }
+  let issuer = pnp.getPackageInformation(locator).packageLocation;
+  const remainder = pathname.substring(issuer.length);
+  const issuers = [issuer].concat(remainder.split(/\/node_modules/));
+
+  issuers[0] += '/';
+  issuer = issuers[0];
+  for (let issuerIdx = 1; issuerIdx < issuers.length; issuerIdx++) {
+    issuers[issuerIdx] = issuers[issuerIdx].substring(1);
+  }
+
+  try {
+    for (let issuerIdx = 1; issuerIdx < issuers.length; issuerIdx++) {
+      if (issuers[issuerIdx] === '') continue;
+      if (issuers[issuerIdx] === '@types') {
+        return issuer;
+      }
+      const nextIssuer = pnp.resolveToUnqualified(issuers[issuerIdx], issuer);
+      if (!nextIssuer) {
+        return undefined;
+      }
+      issuer = nextIssuer + '/';
+    }
+    return issuer;
+  } catch (e) {
+  }
+}
+
+const pnpApiRoots: { [pnpRootDir: string]: any } = {};
+const nonPnpPaths: { [pnpRootDir: string]: boolean } = {};
+
+function getPnpApiForPath(pathname: string) {
+  const basePath = pathname.substring(0, pathname.indexOf('node_modules') - 1);
+  if (nonPnpPaths[basePath]) {
+    return undefined;
+  }
+  for (let knownPnpRootDir of Object.keys(pnpApiRoots)) {
+    if (basePath.startsWith(knownPnpRootDir)) {
+      return pnpApiRoots[knownPnpRootDir];
```

Comment:
Are you 100% sure about the single instance for multiple projects? That seems weird to me, because multiple projects can use different versions of TypeScript and if there was a single instance it already would be broken. That would seem a major oversight.

Reply:
I'm not sure I follow. What do you mean "single instance for multiple projects"? Here one instance of pnp api used for all the paths starting with the same root directory having .pnp.js file. But if the root directory is different, different pnp api will be used.

Reply:
I wanted to ask if the situation `/home/user/project1/.pnp.js/subproject2/.pnp.js` is possible with pnp? If yes, then this logic should certainly be changed, but if not, it should be okay I think